### PR TITLE
deployment: retry failed watch when observed controller has old resource version

### PR DESCRIPTION
**Pull request by mfojtik**
Fixes: https://github.com/openshift/origin/issues/12905

This is pretty hard to reproduce or wrap with unit test :-(

@kargakis PTAL

### CODE REVIEW

**Comment by mfojtik** [comment]
[test]

**Comment by mfojtik** [comment]
[test]

flake: https://github.com/openshift/origin/issues/12911

File: pkg/deploy/registry/rest.go [hunk 1]
```diff
@@ -6,17 +6,20 @@ import (
...
 
+	"github.com/golang/glog"
 	"github.com/openshift/origin/pkg/deploy/api"
 	deployutil "github.com/openshift/origin/pkg/deploy/util"
 )
 
 var (
 	// ErrUnknownDeploymentPhase is returned for WaitForRunningDeployment if an unknown phase is returned.
 	ErrUnknownDeploymentPhase = errors.New("unknown deployment phase")
+	ErrTooOldResourceVersion  = errors.New("too old resource version")
```
Comment (kargakis) [inline pkg/deploy/registry/rest.go]: Isn't this error typed in the watch library?

Comment (mfojtik) [inline pkg/deploy/registry/rest.go]: i will look

Comment (kargakis) [inline pkg/deploy/registry/rest.go]: It's strange that you recreate the error here and return it in place of the previous error which is essentially the same error?

Comment (mfojtik) [inline pkg/deploy/registry/rest.go]: @kargakis nope, I don't see any.

Comment (mfojtik) [inline pkg/deploy/registry/rest.go]: @kargakis well I can return the original error, but I will have to type cast it twice :)

**Comment by mfojtik** [comment]
@smarterclayton @ncdc I've seen similar BZ about lifecycle hooks (same error message). Does something changed in how the watches work in last ~2 weeks?

**Comment by kargakis** [comment]
If the lifecycle hook bug is the one with the mysql pod, I didn't find any hooks on that DC so we may want to check why we used the watch in the first place.

**Comment by mfojtik** [comment]
@kargakis yes, that is what I'm looking at right now. (however, I don't think we touched the lifecycle hooks recently, so I'm interested if something changed in watches that causes this errors to surface now).

**Comment by mfojtik** [comment]
@kargakis you ok merging this?

**Comment by kargakis** [comment]
Can't we at least unit test this? I am scared of untested api changes going in.

**Comment by mfojtik** [comment]
@kargakis agree, unit test sounds reasonable.

**Comment by kargakis** [comment]
@openshift/api-review since this is essentially an api change

**Comment by mfojtik** [comment]
@kargakis unit test added, PTAL

**Comment by smarterclayton** [comment]
API change lgtm

**Comment by kargakis** [comment]
lgtm [merge]

**Pull request merged**